File: iosApp/iosApp/Source/UI/NoteDetail/NoteDetailViewModel.swift
import Foundation
import shared

@MainActor
final class NoteDetailViewModel: ObservableObject {
    
    @Published var noteTtitle: String = ""
    @Published var noteContent: String = ""
    @Published private(set) var noteColor: Int64 = Note.Companion().generateRandomColor()
    @Published private(set) var error: NoteDetailError?
    @Published var isErrorPresented: Bool = false
    @Published private(set) var isSaved: Bool = false
    
    private let noteId: Int64?
    private let noteDataSource: NoteDataSource
    
    init(
        noteId: Int64?,
        noteDataSource: NoteDataSource
    ) {
        self.noteId = noteId
        self.noteDataSource = noteDataSource
    }
    
    func loadNote() async {
        do {
            guard let noteId, let note = try await noteDataSource.getNoteById(id: noteId) else {
                return
            }
            self.noteTtitle = note.title
            self.noteContent = note.content
            self.noteColor = note.colorHex
        } catch {
            showError(NoteDetailError.loadNoteFailed(error))
        }
    }
    
    func saveNote() {
        Task {
            do {
                let id: KotlinLong?
                
                if let noteId {
                    id = KotlinLong(longLong: noteId)
                } else {
                    id = nil
                }
                try await noteDataSource.insertNote(
                    note: Note(
                        id: id,
                        title: noteTtitle,
                        content: noteContent,
                        colorHex: noteColor,
                        created: DateTimeUtil().now()
                    )
                )
                isSaved = true
            } catch {
                showError(NoteDetailError.saveNoteFailed(error))
            }
        }
    }
    
    private func showError(_ error: NoteDetailError) {
        self.error = error
        self.isErrorPresented = true
    }
}

File: androidApp/src/main/java/pl/mietelski/robert/notesapp/android/noteDetail/NoteDetailViewModel.kt
package pl.mietelski.robert.notesapp.android.noteDetail

import androidx.lifecycle.SavedStateHandle
import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import dagger.hilt.android.lifecycle.HiltViewModel
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.SharingStarted
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.combine
import kotlinx.coroutines.flow.stateIn
import kotlinx.coroutines.launch
import pl.mietelski.robert.notesapp.domain.note.Note
import pl.mietelski.robert.notesapp.domain.note.NoteDataSource
import pl.mietelski.robert.notesapp.domain.time.DateTimeUtil
import javax.inject.Inject

@HiltViewModel
class NoteDetailViewModel @Inject constructor(
    private val noteDataSource: NoteDataSource,
    private val savedStateHandle: SavedStateHandle
): ViewModel() {

    private val noteTile = savedStateHandle.getStateFlow("noteTile", "")
    private val isNoteTitleFocused = savedStateHandle.getStateFlow("isNoteTitleFocused", false)
    private val noteContent = savedStateHandle.getStateFlow("noteContent", "")
    private val isNoteContentFocused = savedStateHandle.getStateFlow("isNoteContentFocused", false)
    private val noteColor = savedStateHandle.getStateFlow("noteColor", Note.generateRandomColor())

    val state = combine(
        noteTile,
        isNoteTitleFocused,
        noteContent,
        isNoteContentFocused,
        noteColor
    ) { title, isTitleFocused, content, isContentFocused, color ->
        NoteDetailState(
            noteTile = title,
            isNoteTitleHintVisible = title.isEmpty() && !isTitleFocused,
            noteContent = content,
            isNoteContentHintVisible = content.isEmpty() && !isContentFocused,
            noteColor = color
        )
    }.stateIn(viewModelScope, SharingStarted.WhileSubscribed(5000), NoteDetailState())

    private val _hasNoteBeenSaved = MutableStateFlow(false)
    val hasNoteBeenSaved = _hasNoteBeenSaved.asStateFlow()

    private var existsNoteId: Long? = null

    init {
        savedStateHandle.get<Long>("noteId")?.let { noteId ->
            if (noteId == -1L) {
                return@let
            }
            this.existsNoteId = noteId
            viewModelScope.launch {
                noteDataSource.getNoteById(noteId)?.let { note ->
                    savedStateHandle["noteTile"] = note.title
                    savedStateHandle["noteContent"] = note.content
                    savedStateHandle["noteColor"] = note.colorHex
                }
            }
        }
    }

    fun onNoteTileChange(text: String) {
        savedStateHandle["noteTile"] = text
    }

    fun onNoteTitleFocusedChange(isFocused: Boolean) {
        savedStateHandle["isNoteTitleFocused"] = isFocused
    }

    fun onNoteContentChange(text: String) {
        savedStateHandle["noteContent"] = text
    }

    fun onNoteContentFocusedChange(isFocused: Boolean) {
        savedStateHandle["isNoteContentFocused"] = isFocused
    }

    fun saveNote() {
        viewModelScope.launch {
            noteDataSource.insertNote(
                Note(
                    id = existsNoteId,
                    title = noteTile.value,
                    content = noteContent.value,
                    colorHex = noteColor.value,
                    created = DateTimeUtil.now()
                )
            )
            _hasNoteBeenSaved.value = true
        }
    }
}